server/src/main/java/io/crate/execution/engine/aggregation/impl/util/BigDecimalValueWrapper.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.engine.aggregation.impl.util;
+
+import java.math.BigDecimal;
+
+public final class BigDecimalValueWrapper implements NumericValueHolder {
+
+    private BigDecimal value;
+    private boolean hasValue;
+
+    public BigDecimalValueWrapper(BigDecimal value) {
+        this.value = value;
+    }
+
+    @Override
+    public BigDecimal value() {
+        return value;
+    }
+
+    public boolean hasValue() {
+        return hasValue;
+    }
+
+    public void setValue(BigDecimal value) {
```

Comment:
Not sure but maybe even better to move the `add` and `subtract` logic here instead of exposing the `setValue`? ` OverflowAwareMutableLong`  already implements `add`...

Reply:
Was thinking about that - I already changed regular AverageState to such implementation - https://github.com/crate/crate/pull/11616/files
but here my intention was to keep `OverflowAwareMutableLong.add`accepting primitive long but not a Number (suitable for Long and BigDecimal). Is it way if I make add accept Number and add a bit of auto boxing?

Reply:
I think it worth to move add, subtract and reduce to `NumericAverageState`  - thanks for the suggestion!  This way I make it consistent with [another PR](https://github.com/crate/crate/pull/11616/files) for regular average aggregation (don't remember why I didn't make it consistent in the first place, probably my first draft changed OverflowAwareMutableLong too much and I didn't like it and on the final version I forgot about it).

I still need to have add in BigDecimalValueWrapper - but `subtract` is redundant as it's just add(-number) and count is handled in NumericAverageState itself.

```
 public void add(Number value) {
        // value is either BigDecimal or long
        sum.add() //BigDecimalValueWrapper needs that method, OAML already has it
        count++;
    }

    public void subtract(Number value) {
        sum.add(-value)
        count--;
    }

    public NumericAverageState reduce(@Nonnull NumericAverageState state) {
        this.sum.add(state.sum.value()) //Here  OAML would need to have add(BigDecimal) method
        this.count += state.count;
    }
```
 
the only concern is `OverflowAwareMutableLong.add` is currently working with primitive long to avoid auto boxing thingy- now I would introduce add(Number number) into NumericValueHolder (or probably better name AverageStateNominator) and make both OverflowAwareMutableLong and BigDecimalValueWrapper implement it - in case of OverflowAwareMutableLong it might be that I need another add to accept Bigdecimal because of 
` this.sum.add(state.sum.value()) //Here  OAML would need to have add(BigDecimal) method`

Reply:
As discussed the `OverflowAwareMutableLong.add(long)` impl to avoid boxing/unboxing, which I wasn't really aware of, completely screws my idea :(
I agree that optimization here is better than nice abstraction. 
So I have no better idea currently than leave it as it.